Amendments for equity products

minesh-s-patel opened the pull request:
See Issue #3087

dshoneisda reviewed (approved):
Approved as per DPBE WG 14th Aug

chrisisla commented on rosetta-source/src/main/rosetta/product-qualification-func.rosetta:
```diff
@@ -681,6 +687,54 @@ func Qualify_BaseProduct_EquityForward: <"Qualifies a product as having the Asse
+        Qualify_AssetClass_Equity(economicTerms) = True
+                // qualifies that only the option payout exist and all other payouts are absent
+            and economicTerms -> payout -> forwardPayout only exists
+                // qualifies that the underlier is an equity security
```
isn't this qualifying an index rather than an equity?

minesh-s-patel replied:
Looks OK to me - logic is. It qualifies when all of the below are true:

1. Qualify_AssetClass_Equity (its an equity)
2. has 1 `forwardPayout`
3. the `forwardPayout` has an `underlier` that is an `index`

chrisisla commented on rosetta-source/src/main/rosetta/product-qualification-func.rosetta:
```diff
@@ -681,6 +687,54 @@ func Qualify_BaseProduct_EquityForward: <"Qualifies a product as having the Asse
+            and economicTerms -> payout -> forwardPayout only exists
+                // qualifies that the underlier is an equity security
+            and economicTerms -> payout -> forwardPayout only-element -> underlier -> index exists
+                // qualifies that an option style is defined and that no feature other than averaging exists
```
same as previous query on the relevance of this comment

chrisisla commented on rosetta-source/src/main/rosetta/product-qualification-func.rosetta:
```diff
@@ -681,6 +687,54 @@ func Qualify_BaseProduct_EquityForward: <"Qualifies a product as having the Asse
+            and economicTerms -> payout -> forwardPayout only-element -> underlier -> index exists
+                // qualifies that an option style is defined and that no feature other than averaging exists
+ 
+func Qualify_EquityForward_PriceReturnBasicPerformance_Basket: <"Qualifies a product as a plain vanilla Equity Option.  The determination of the qualification is based on the economic terms and the following criteria: 1) An option product  for which the underlier is a single stock and 2) No special option feature exists other than option averaging."> //
```
// again and looking for basket not a stock

chrisisla commented on rosetta-source/src/main/rosetta/product-qualification-func.rosetta:
```diff
@@ -681,6 +687,54 @@ func Qualify_BaseProduct_EquityForward: <"Qualifies a product as having the Asse
+        Qualify_AssetClass_Equity(economicTerms) = True
+                // qualifies that only the option payout exist and all other payouts are absent
+            and economicTerms -> payout -> forwardPayout only exists
+                // qualifies that the underlier is an equity security
```
it's looking for a basket not a security

chrisisla commented on rosetta-source/src/main/rosetta/product-qualification-func.rosetta:
```diff
@@ -681,6 +687,54 @@ func Qualify_BaseProduct_EquityForward: <"Qualifies a product as having the Asse
+            and economicTerms -> payout -> forwardPayout only exists
+                // qualifies that the underlier is an equity security
+            and economicTerms -> payout -> forwardPayout only-element -> underlier -> basket exists
+                // qualifies that an option style is defined and that no feature other than averaging exists
```
again don't think we need this comment here

chrisisla commented on RELEASE.md:
```diff
@@ -1,13 +1,23 @@
 _What is being released?_
 
-This release updates the `DSL` dependency.
+This issue was discusses at the CDM Derivatives Products and Business Events Working Group on August 14th, 2024 #3077 and resolves items 1, 2 and 4 from Issue #3087. Item 3 was discussed and will be contributed at a later time.
```
discussed ;)

minesh-s-patel commented on rosetta-source/src/main/rosetta/product-qualification-func.rosetta:
```diff
@@ -681,6 +687,54 @@ func Qualify_BaseProduct_EquityForward: <"Qualifies a product as having the Asse
         Qualify_AssetClass_Equity(economicTerms) = True
             and economicTerms -> payout -> forwardPayout only exists
 
+func Qualify_EquityForward_PriceReturnBasicPerformance_SingleName: <"Qualifies a product as a plain vanilla Equity Option.  The determination of the qualification is based on the economic terms and the following criteria: 1) An option product  for which the underlier is a single stock and 2) No special option feature exists other than option averaging."> //
```
👍

minesh-s-patel commented on rosetta-source/src/main/rosetta/product-qualification-func.rosetta:
```diff
@@ -681,6 +687,54 @@ func Qualify_BaseProduct_EquityForward: <"Qualifies a product as having the Asse
+            and economicTerms -> payout -> forwardPayout only exists
+                // qualifies that the underlier is an equity security
+            and economicTerms -> payout -> forwardPayout only-element -> underlier -> security exists
+                // qualifies that an option style is defined and that no feature other than averaging exists
```
I do not think that these (incorrect) comments add anything here. I opt to remove them.

minesh-s-patel commented on rosetta-source/src/main/rosetta/product-qualification-func.rosetta:
```diff
@@ -274,10 +274,16 @@ func Qualify_UnderlierProduct_Equity: <"Qualifies a product as having the Asset
             or underlier -> security -> securityType = SecurityTypeEnum -> Warrant
+            or (underlier -> security -> securityType = SecurityTypeEnum -> Fund
+                and underlier -> security -> fundType = FundProductTypeEnum -> OtherFund)
+            or underlier -> security -> securityType = SecurityTypeEnum -> Warrant
```
It is `FundProductTypeEnum -> MutualFund` vs `FundProductTypeEnum -> OtherFund`.

Arguably - the if statement can be changed to an `or`, but I think is better to have less parentheses in these long statements to aid readability.

chrisisla replied:
sorry, I probably wasn't very clear, its the `securityType = Warrant` line that I think is the duplicate.

Based upon the brackets it looks like the logic (abbreviated for clarity) is:

```
if securityType = Equity
or securityType = Fund and fundType = ExchangeTradedFund
or securityType = Fund and fundType = MutualFund
or securityType = Warrant
or securityType = Fund and fundType = OtherFund
or securityType = Warrant
or securityType = Certificate
```
So we either don't need the first check for a Warrant (if we're just checking for a securityType of Warrant regardless of the fundType), or we possibly need to move the brackets?

Hopefully that is clearer...!

minesh-s-patel replied:
Thanks - got it.

1. I removed the duplicated `underlier -> security -> securityType = SecurityTypeEnum -> Warrant`.
2. I updated the qualification function definitions to be inline with what they do (previously - lazily copied and pasted).

For the record - I did not make these changes - I am just contributing them ;).

minesh-s-patel reviewed (commented):
@chrisisla 

I resolved/answered your questions. Can you please re-review and hit the approve button if all is OK?

Really appreciate the review! I better up my game.

chrisisla commented on rosetta-source/src/main/rosetta/product-qualification-func.rosetta:
```diff
@@ -681,6 +687,45 @@ func Qualify_BaseProduct_EquityForward: <"Qualifies a product as having the Asse
+            and economicTerms -> payout -> forwardPayout only exists
+            and economicTerms -> payout -> forwardPayout only-element -> underlier -> security exists
+
+func Qualify_EquityForward_PriceReturnBasicPerformance_SingleIndex: <"Qualifies a product as a plain vanilla Equity Option.  The determination of the qualification is based on the economic terms and the following criteria: 1) An option product  for which the underlier is a single stock and 2) No special option feature exists other than option averaging.">
```
I don't think I was very clear on this either 😅 The function is fine, it's qualifying where you have an index whereas the main comment states it is for a single stock. I think all that needs to be done is to update the comment to reference "an index" rather than "single stock"

chrisisla commented on rosetta-source/src/main/rosetta/product-qualification-func.rosetta:
```diff
@@ -681,6 +687,45 @@ func Qualify_BaseProduct_EquityForward: <"Qualifies a product as having the Asse
+            and economicTerms -> payout -> forwardPayout only exists
+            and economicTerms -> payout -> forwardPayout only-element -> underlier -> index exists
+
+func Qualify_EquityForward_PriceReturnBasicPerformance_Basket: <"Qualifies a product as a plain vanilla Equity Option.  The determination of the qualification is based on the economic terms and the following criteria: 1) An option product  for which the underlier is a single stock and 2) No special option feature exists other than option averaging.">
```
And again on this one, we're looking for a basket so we just need to update the function comment to read "basket" rather than "single stock" 🙂

chrisisla commented:
> @chrisisla
> 
> I resolved/answered your questions. Can you please re-review and hit the approve button if all is OK?
> 
> Really appreciate the review! I better up my game.

No problem! I don't think my comments were very clear about what might need to be updated... sorry! Can you just double check the logic in the securityType/fundType "or" sequence please, I think we can remove one of the checks for a Warrant.

Thanks!

chrisisla reviewed (dismissed):
Awesome! Thanks @minesh-s-patel 👍

minesh-s-patel commented:
> Awesome! Thanks @minesh-s-patel 👍

Can you please hit the approve button? (I can't approve my own changes - GitHub says NO). :)

chrisisla reviewed (changes requested):
Release.md seems to have changed, it has the old details in it again too?

minesh-s-patel commented:
> Release.md seems to have changed, it has the old details in it again too?

Hi @chrisisla - This is OK. The release md has all the changes since the last release (including the dependency update).

chrisisla commented:
> > Release.md seems to have changed, it has the old details in it again too?
> 
> Hi @chrisisla - This is OK. The release md has all the changes since the last release (including the dependency update).

Ah okay, thanks for the explanation Minesh, that makes more sense now 🙂
